src/drive/web/modules/navigation/duck/actions.jsx

```diff
@@ -505,13 +510,49 @@ export const openFileWith = (id, filename) => {
   }
 }
 
-export const deriveEncryptionKey = passphrase => {
+export const decryptVaultEncryptionKey = passphrase => {
   return async (dispatch, _, { client }) => {
+    // TODO derive secret key + get encrypted vault key + decrypt it
     const salt = client.getStackClient().uri
     const key = await deriveKey(passphrase, salt)
     return dispatch({
-      type: DERIVE_ENCRYPTION_KEY,
+      type: DECRYPT_VAULT_ENCRYPTION_KEY,
       key
     })
   }
 }
+
+export const createVaultEncryptionKey = passphrase => {
+  return async (dispatch, _, { client }) => {
+    // Derive secret key
+    const stackClient = client.getStackClient()
+    const salt = stackClient.uri
+    const secretKey = await deriveKey(passphrase, salt)
+    // Generate vault key and wrap it
+    const vaultKey = await generateAESVaultKey({ algorithm: 'AES-KW' })
+    // TODO: find a suitable kid
+    const jwe = await wrapAESKey(vaultKey, secretKey, '123')
+
+    // Save the wrapped key in settings
+    const settings = await client.query(
+      client.find('io.cozy.settings').getById('io.cozy.settings.instance')
+    )
+    let encryption
+    if (settings.data.encryption && settings.data.encryption.keys) {
+      settings.data.encryption.keys.push(jwe)
+      const keys = settings.data.encryption.keys
+      encryption = { ...settings.data.encryption, keys }
```

Comment:
I don't understand this line : it looks like `encryption` will embed 2 copies of `keys` ..?

Reply:
The spread operator is used to "update" the `settings.data.encryption` object with the newly created `keys` array, without erasing other fields (if any)  in `settings.data.encryption`